[Fix-10181] Fix the logic of judging that the tenant does not exist

lugela opened the pull request:
[Fix-10181] Fix the logic of judging that the tenant does not exist
Use the linux command as id to get the user information that exists in /etc/passwd file and the cached sssd user.
for example:
id test
1. exist in /etc/passwd file or ldap :  uid=1030(test) gid=1030(test) groups=1030(test)
2. no exist  in /etc/passwd file and ldap: id: test: no such user

Temporarily unable to test the system for windows and mac

Commit:
Merge pull request #1 from apache/dev

同步

Commit:
[Fix-10181] Fix the logic of judging that the tenant does not exist

Use the linux command as id to get the user information that exists in /etc/passwd file and the cached sssd user.
for example:
id test
1. exist in /etc/passwd file or ldap :  uid=1030(test) gid=1030(test) groups=1030(test)
2. no exist  in /etc/passwd file and ldap: id: test: no such user

Temporarily unable to test the system for windows and mac

songjianet commented:
205's question needs to be cut out of 205 and merged back into 205. If the dev branch has the same problem, it needs to be modified in dev. @lugela

Commit:
[Fix-10181] Fix the logic of judging that the tenant does not exist

Use the linux command as id to get the user information that exists in /etc/passwd file and the cached sssd user.
for example:
id test
1. exist in /etc/passwd file or ldap :  uid=1030(test) gid=1030(test) groups=1030(test)
2. no exist  in /etc/passwd file and ldap: id: test: no such user

Temporarily unable to test the system for windows and mac

Commit:
[Fix-10181] Fix the logic of judging that the tenant does not exist

Use the linux command as id to get the user information that exists in /etc/passwd file and the cached sssd user.
for example:
id test
1. exist in /etc/passwd file or ldap :  uid=1030(test) gid=1030(test) groups=1030(test)
2. no exist  in /etc/passwd file and ldap: id: test: no such user

Temporarily unable to test the system for windows and mac

lugela commented:
both version 2.0.5 and dev have this bug. In this case, do you need to mention 2.0.5 or dev, if it is 2.0.5 is 2.0.5-release?@songjianet

lugela commented:
seeing that the pr has been waiting for status to be reported, I want to ask if the administrator needs to manually click？

lugela commented:
@songjianet Hi, can you answer the above question?

songjianet commented:
> @songjianet Hi, can you answer the above question?

Repair the dev handle first. The maintenance of 3.x will be greater than that of 2.x. You need to cut the branch from dev, merge it back into dev, and then cherry-pick it to the 300bp branch

lugela commented:
> > @songjianet Hi, can you answer the above question?
> 
> Repair the dev handle first. The maintenance of 3.x will be greater than that of 2.x. You need to cut the branch from dev, merge it back into dev, and then cherry-pick it to the 300bp branch

ok,I found that the above SonarCloud  failed, but I didn't understand the reason for the failure after clicking it. Can you guide me on what to do?

caishunfeng commented on dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/OSUtils.java:
```diff
@@ -156,65 +148,75 @@ public static double cpuUsage() {
+        try{
+            String result = exeCmd("id "+ tenantCode);
+            if (!StringUtils.isEmpty(result)){
+                return result.contains("uid=");
```
![image](https://user-images.githubusercontent.com/11962619/169763464-52adaf81-167b-4a42-b5be-77afa60dfa05.png)
I tried it locally and found that there may be a problem with this judgment

caishunfeng commented on dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/OSUtils.java:
```diff
@@ -156,65 +148,75 @@ public static double cpuUsage() {
+        try{
+            String result = exeCmd("id "+ tenantCode);
+            if (!StringUtils.isEmpty(result)){
+                return result.contains("uid=");
```
![image](https://user-images.githubusercontent.com/11962619/169764705-34e0afab-86ec-4259-ad35-c4ccaebf55f9.png)

I tried it locally and found that there may be a problem with this judgment

lugela replied:
Hi, do you have any suggestions here? Because the user information in ldap will not exist in the /etc/passwd file

caishunfeng replied:
> ![image](https://user-images.githubusercontent.com/11962619/169764705-34e0afab-86ec-4259-ad35-c4ccaebf55f9.png)
> 
> I tried it locally and found that there may be a problem with this judgment

Sorry, I read it wrong.

SbloodyS commented on dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/OSUtils.java:
```diff
@@ -156,65 +148,75 @@ public static double cpuUsage() {
-                }
+    private static boolean existTenantCodeInLinux(String tenantCode) throws IOException {
+        try{
+            String result = exeCmd("id "+ tenantCode);
```
This command ```id ${tennantCode}``` may not be compatiable with users environment. It may throw ```permission denied``` exception when the user ```dolphinscheduler``` did not have execute permissions.

SbloodyS replied:
I think we can provide a switch param in ```common.properties``` to determine whether to use distributed authentication. If it is closed, leave it as it is. If it's ```on```, then use this function. And also we should put this change in the document so that users can chose it according to their scene. What do you think?

lugela replied:
> This command `id ${tennantCode}` may not be compatiable with users environment. It may throw `permission denied` exception when the user `dolphinscheduler` did not have execute permissions.

At present, it is verified that the id ${tennantCode} has permission to execute under any user, and does not need sudo

lugela replied:
> I think we can provide a switch param in `common.properties` to determine whether to use distributed authentication. If it is closed, leave it as it is. If it's `on`, then use this function. And also we should put this change in the document so that users can chose it according to their scene. What do you think?

This is a good idea because of the compatibility of the id command I can't test if all liunx versions are supported。

lugela replied:
> I think we can provide a switch param in `common.properties` to determine whether to use distributed authentication. If it is closed, leave it as it is. If it's `on`, then use this function. And also we should put this change in the document so that users can chose it according to their scene. What do you think?

hi, have time to see if the current logic is OK?

SbloodyS replied:
> > I think we can provide a switch param in `common.properties` to determine whether to use distributed authentication. If it is closed, leave it as it is. If it's `on`, then use this function. And also we should put this change in the document so that users can chose it according to their scene. What do you think?
> 
> hi, have time to see if the current logic is OK?

I will take a look later.

Commit:
Merge pull request #2 from apache/dev

更新

Commit:
Merge pull request #3 from lugela/dev

更新dev

Commit:
[Fix-10181] Fix the logic of judging that the tenant does not exist

The configuration item adds 'tenant-distributed-user' in worker application.yaml to make it suitable for distributed users. If it is false, the original logic remains unchanged.

At present, considering that it is a distributed user, it should not be allowed to create users in linux

Use the linux command as id to get the user information that exists in /etc/passwd file and the cached sssd user.
for example:
id test
1. exist in /etc/passwd file or ldap :  uid=1030(test) gid=1030(test) groups=1030(test)
2. no exist  in /etc/passwd file and ldap: id: test: no such user

Temporarily unable to test the system for windows and mac

Commit:
[Fix-10181] Fix the logic of judging that the tenant does not exist

Add test method

The configuration item adds 'tenant-distributed-user' in worker application.yaml to make it suitable for distributed users. If it is false, the original logic remains unchanged.

At present, considering that it is a distributed user, it should not be allowed to create users in linux

Use the linux command as id to get the user information that exists in /etc/passwd file and the cached sssd user.
for example:
id test
1. exist in /etc/passwd file or ldap :  uid=1030(test) gid=1030(test) groups=1030(test)
2. no exist  in /etc/passwd file and ldap: id: test: no such user

Temporarily unable to test the system for windows and mac

SbloodyS commented:
Hi @lugela , please also add this to the docs.
```docs/docs/en/architecture/configuration.md```
```docs/docs/zh/architecture/configuration.md```

Commit:
[Fix-10181] Fix the logic of judging that the tenant does not exist

Add parameter description to configuration.md

Add test method

The configuration item adds 'tenant-distributed-user' in worker application.yaml to make it suitable for distributed users. If it is false, the original logic remains unchanged.

At present, considering that it is a distributed user, it should not be allowed to create users in linux

Use the linux command as id to get the user information that exists in /etc/passwd file and the cached sssd user.
for example:
id test
1. exist in /etc/passwd file or ldap :  uid=1030(test) gid=1030(test) groups=1030(test)
2. no exist  in /etc/passwd file and ldap: id: test: no such user

Temporarily unable to test the system for windows and mac

lugela commented:
> Hi @lugela , please also add this to the docs. `docs/docs/en/architecture/configuration.md` `docs/docs/zh/architecture/configuration.md`

OK，Commit has been added

Commit:
[Fix-10181] Fix the logic of judging that the tenant does not exist

Add parameter description to configuration.md

Add test method

The configuration item adds 'tenant-distributed-user' in worker application.yaml to make it suitable for distributed users. If it is false, the original logic remains unchanged.

At present, considering that it is a distributed user, it should not be allowed to create users in linux

Use the linux command as id to get the user information that exists in /etc/passwd file and the cached sssd user.
for example:
id test
1. exist in /etc/passwd file or ldap :  uid=1030(test) gid=1030(test) groups=1030(test)
2. no exist  in /etc/passwd file and ldap: id: test: no such user

Temporarily unable to test the system for windows and mac

SbloodyS commented on docs/docs/en/architecture/configuration.md:
```diff
@@ -246,7 +246,8 @@ worker.heartbeat.interval|10|worker-service heartbeat interval, the unit is seco
 worker.reserved.memory|0.3|worker reserved memory, only lower than system available memory, worker server can be dispatched tasks. default value 0.3, the unit is G
 worker.groups|default|worker groups separated by comma, e.g., 'worker.groups=default,test' <br> worker will join corresponding group according to this config when startup
-
+worker.tenant.auto.create|true|The tenant is worker's user, which is used by the worker to submit the job. If system does not have this user, it will be automatically created after the parameter is true.
```
Is it better to use ```it will be automatically created after the parameter worker.tenant.distributed.user is true.``` instead of ```it will be automatically created after the parameter is true```?

SbloodyS commented on docs/docs/zh/architecture/configuration.md:
```diff
@@ -235,6 +235,8 @@ worker.heartbeat.interval|10|worker心跳间隔,单位为秒
 worker.max.cpuload.avg|-1|worker最大cpuload均值,只有高于系统cpuload均值时,worker服务才能被派发任务. 默认值为-1: cpu cores * 2
 worker.reserved.memory|0.3|worker预留内存,只有低于系统可用内存时,worker服务才能被派发任务,单位为G
 worker.groups|default|worker分组配置,逗号分隔,例如'worker.groups=default,test' <br> worker启动时会根据该配置自动加入对应的分组
+worker.tenant.auto.create|true|租户对应的是worker系统的用户,用于worker提交作业所使用的用户.如果系统中没有这个用户,开启后将会自动创建
```
Same as english's doc.

SbloodyS commented on dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/OSUtils.java:
```diff
@@ -249,6 +249,36 @@ private static List<String> getUserListFromWindows() throws IOException {
+            }
+        }catch (Exception e){
+            String message = e.getMessage();
+            //Because ShellExecutor method throws  exception to the linux return status is not 0
```
Please remove duplicated space in ```throws exception```

SbloodyS commented on dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/OSUtils.java:
```diff
@@ -249,6 +249,36 @@ private static List<String> getUserListFromWindows() throws IOException {
+            //Because ShellExecutor method throws  exception to the linux return status is not 0
+            //not exist user return status is 1
+            if (message.contains("no such user")){
+                logger.warn(message);
```
Is it better to use error instead of warn? Becasue the worker would not execute any task without user.

SbloodyS commented on dolphinscheduler-standalone-server/src/main/resources/application.yaml:
```diff
@@ -125,6 +125,9 @@ worker:
   host-weight: 100
   # worker tenant auto create
   tenant-auto-create: true
+  # worker tenant is distributed user. If this is true, tenant-auto-create has no effect. Will automatically create users without creating user.
```
Same as docs.

SbloodyS commented on dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/processor/TaskExecuteProcessor.java:
```diff
@@ -116,12 +117,21 @@ public void process(Channel channel, Command command) {
+            //create tenants,so TenantAutoCreate has no effect
+            if (workerConfig.isTenantDistributedUser() && SystemUtils.IS_OS_LINUX){
+                osUserExistFlag = OSUtils.existTenantCodeInLinux(taskExecutionContext.getTenantCode());
+            }else {
```
It's recommand to use ```else if``` instead of ```else{} if() {}````

SbloodyS commented on dolphinscheduler-worker/src/main/resources/application.yaml:
```diff
@@ -64,6 +64,9 @@ worker:
   host-weight: 100
   # worker tenant auto create
   tenant-auto-create: true
+  # worker tenant is distributed user. If this is true, tenant-auto-create has no effect. Will automatically create users without creating user.
```
Same as docs.

Commit:
[Fix-10181] Fix the logic of judging that the tenant does not exist

Add parameter description to configuration.md

Add test method

The configuration item adds 'tenant-distributed-user' in worker application.yaml to make it suitable for distributed users. If it is false, the original logic remains unchanged.

At present, considering that it is a distributed user, it should not be allowed to create users in linux

Use the linux command as id to get the user information that exists in /etc/passwd file and the cached sssd user.
for example:
id test
1. exist in /etc/passwd file or ldap :  uid=1030(test) gid=1030(test) groups=1030(test)
2. no exist  in /etc/passwd file and ldap: id: test: no such user

Temporarily unable to test the system for windows and mac

SbloodyS reviewed (approved):
LGTM

caishunfeng reviewed (approved):
+1
